base/error.jl

```diff
@@ -208,6 +202,7 @@ julia> @assert isodd(3) "What even are numbers?"
 ```
 """
 macro assert(ex, msgs...)
+    @isdefined(isdebug) && !(isdebug()) && return nothing
```

Comment:
This is for the `Compiler` module where `isdebug` is not defined.

Reply:
options.jl is already included by this point during bootstrap. So you could just eliminate the `julia_debug` global and define `isdebug()` directly in error.jl as

```julia
isdebug() = JLOptions().debug_level >= 2
```

?

Reply:
Yeah, but that ties `isdebug` directly to the debugging level (while it is not only tied to it in the sense that it is used to set the variable on startup). With the current implementation you can force julia debug mode without having to start julia itself with `-g2`. I don't know if the extra flexibility is worth it though.

Reply:
Allowing `julia_debug` to change dynamically seems like it could be confusing and lead to inconsistencies between the precompiled code in different modules which are loaded into the session.

But maybe that's a reasonable tradeoff for flexibility, I'm not sure.

Example crazy thing:

```julia
module A

if Base.isdebug()
    struct X
        x
        debug_field
    end
else
    struct X
        x
    end
end

end
```

Now other modules which depend on `A` can't depend on the value of `isdebug()` to use `X` correctly (including macros defined by A and expanded in another module).

Reply:
> Allowing julia_debug to change dynamically seems like it could be confusing and lead to inconsistencies between the precompiled code in different modules which are loaded into the session.

Yeah, but `julia_debug` is internal so we'll just not change it dynamically in a way like that :P. But yes, I did think about changing this to exactly what you say. Can be discussed at triage as well. This problem already kind of exists in changing environments dynamically, by the way.